samples/chat-demo/main.go

```diff
@@ -97,33 +110,45 @@ func run() error {
 
 	fmt.Println("Initializing...")
 
-	ownID, err := strconv.Atoi(string(args.OwnID))
+	// ================================================================================
+	// Load system membership info: IDs, addresses, ports, etc...
+	// ================================================================================
+
+	ownNumericID, err := strconv.Atoi(string(args.OwnID))
 	if err != nil {
-		return fmt.Errorf("unable to convert node ID: %w", err)
-	}
-	if ownID < 0 || ownID > nodeNumber-1 {
-		return fmt.Errorf("ID must be in [0, %d]", nodeNumber-1)
+		return fmt.Errorf("node IDs must be numeric in the sample app: %w", err)
 	}
 
-	// ================================================================================
-	// Generate system membership info: addresses, ports, etc...
-	// ================================================================================
+	initialMembership, err := loadMembership(args.InitMembershipFile)
+	if err != nil {
+		return fmt.Errorf("could not load membership: %w", err)
+	}
+	if err := args.InitMembershipFile.Close(); err != nil {
+		return fmt.Errorf("could not close membership file: %w", err)
+	}
+	addresses, err := membershipIPs(initialMembership)
+	if err != nil {
+		return fmt.Errorf("could not load node IPs: %w", err)
+	}
 
 	// IDs of nodes that are part of the system.
-	// This example uses a static configuration of nodeNumber nodes.
-	nodeIDs := make([]t.NodeID, nodeNumber)
-	for i := 0; i < nodeNumber; i++ {
-		nodeIDs[i] = t.NewNodeIDFromInt(i)
-	}
+	nodeIDs := maputil.GetKeys(initialMembership)
+	sort.Slice(nodeIDs, func(i, j int) bool {
+		return nodeIDs[i] < nodeIDs[j]
+	})
 
 	// Generate addresses and ports for client request receivers.
 	// Each node uses different ports for receiving protocol messages and requests.
 	// These addresses will be used by the client code to know where to send its requests
 	// (each client sends its requests to all request receivers). Each request receiver,
 	// however, will only submit the received requests to its associated Node.
 	reqReceiverAddrs := make(map[t.NodeID]string)
-	for i := range nodeIDs {
-		reqReceiverAddrs[t.NewNodeIDFromInt(i)] = fmt.Sprintf("127.0.0.1:%d", reqReceiverBasePort+i)
+	for nodeID, nodeIP := range addresses {
+		numericID, err := strconv.Atoi(string(nodeID))
+		if err != nil {
+			return fmt.Errorf("node IDs must be numeric in the sample app: %w", err)
+		}
+		reqReceiverAddrs[nodeID] = fmt.Sprintf("%s:%d", nodeIP, reqReceiverBasePort+numericID)
```

Comment:
use `net.JoinHostPort` instead of fmt.Sprintf

Reply:
Fixed.